README.md
# Turquoise
Universal Discord OAuth Gate

# Requirements
* RethinkDB
* RedisDB
* NodeJS

# Installation


CHANGELOG.md
# CHANGELOG.md

## v1.1.0
- Remove the need for hardcoding urls
- Update copyrights with respective years

## v1.0.1
- Add changelog
- Stop people from authorising multiple times, as this causes a 401 unauthorised error on their end


index.js
/*
 This file is part of Turquoise.
 Copyright (c) Snazzah 2016-2019
 Copyright (c) Yamboy1 (and contributors) 2019

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

(async ()=>{

const express = require('express')
const expressVue = require('express-vue')
const path = require('path')
const config = require('./config.json')
const pkg = require('./package.json')

const app = express()
const Database = require('./utils/db')
const db = new Database()
await db.connect(config.r)
let profiles = await db.getAllProfiles()
const tokengen = require('token-generator')(config.tg)

const fetch = require("node-fetch");

const db2 = new Database();
await db2.connect(config.r2);

    app.get("/trello/post/:token", async (req, res) => {
        console.log("here");
        if (!req.params.token) res.status(400).send("Please give a token")
        const ticket = req.header("authorization");
        console.log(ticket);
        const details = await db.getTicket(ticket);
        if (await db2.getUser(details.user) !== null){
            return res.status(400).send("You're already authorized");
        }
        const result = await (
          fetch(`https://api.trello.com/1/tokens/${req.params.token}?token=${req.params.token}&key=${config.trelloKey}`)
            .then(res => res.json())
            .then(json => json.idMember)
        ).catch(() => new Error("coudn't get trello user id"))
        console.log(result);
        console.log(details);
        db2.addToken(details.user, req.params.token, result);
        res.send('POST request to the homepage')
    });


app.use(expressVue.init({
    rootPath: path.join(__dirname, 'views'),
    vue: {
        head: {
            title: pkg.name,
            meta: [
                { script: 'https://unpkg.com/vue@2.4.2/dist/vue.js' },
                { name: 'viewport', content: 'width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no' },
                { style: '/main.css' },
                { name: 'theme-color', content: '#40E0D0' },
                { rel: 'icon', type: 'image/png', href: '/images/icon.png' },
                { name: 'title', content: pkg.name },
                { name: 'description', content: 'Turquoise OAuth Gate' },
                { property: 'og:title', content: pkg.name },
                { property: 'og:description', content: 'Turquoise OAuth Gate' },
                { name: 'owner', content: 'Snazzah' }
            ]
        }
    },
    data: {
        availableProfiles: profiles.map(p => ({ id: p.id, name: p.name, client_id: p.client_id }))
    }
}));

app.use((req, res, next) => {
    res.displayError = (code = 404, error = 'Not Found') => res.status(code).renderVue('404',
      { error: `${code} - ${error}` },
      { head: { meta: [
        { rel: 'icon', type: 'image/png', href: '/images/icon-error.png' },
        { name: 'theme-color', content: '#ee5253' }
      ] } })
    Object.assign(req, { db, config, pkg, tokengen })
    res.header("Access-Control-Allow-Origin", "*")
    res.header("Access-Control-Allow-Headers", "*")
    return next()
});

app.use(express.static(__dirname+'/public'))
app.use('/connect', require('./routes/connect'))
app.use('/endpoints', require('./routes/endpoints'))
app.use('/trello', require('./routes/trello'))
app.use('/*', require('./routes/404'))

app.listen(config.port, () => {
  console.info(`Running on port ${config.port}`)
})

})()


utils/db.js
/*
 This file is part of Turquoise.
 Copyright (c) Snazzah 2016-2019
 Copyright (c) Yamboy1 (and contributors) 2019

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const rdb = require("rethinkdb")
const { EventEmitter } = require("eventemitter3")

module.exports = class Database extends EventEmitter {
  async connect({host = 'localhost', port, user, password, database}) {
    let err, conn = await rdb.connect({host, port, user, password})
    if(err) return this.onError(err)
    console.log('[DB]', 'Connected')
    if(database) conn.use(database)
    conn.on('close', this.onClose.bind(this));
    conn.on('timeout', this.onTimeout.bind(this));
    this.conn = conn
    this.host = host
    this.port = port
    this.user = user
    this.password = password
    this.database = database
  }

  get r() {
    return rdb
  }

  getProfile(id) {
    return rdb.table('profiles').get(id).run(this.conn)
  }

  getTicket(ticket) {
    return rdb.table('tickets').get(ticket).run(this.conn)
  }

  addTicket(ticket, profile, token, user) {
    return rdb.table('tickets').insert({
      id: ticket, profile, expire: Date.now() + (86400000*3), token, user: user.id, username: user.username + "#" + user.discriminator
    }).run(this.conn)
  }

  addToken(id, trelloToken, trelloID) {
    return rdb.table('users').insert({
      current: null,
      id,
      trelloID,
      trelloToken
  }).run(this.conn);
  }

  getUser(id) {
    return rdb.table('users').get(id).default(null).run(this.conn);
  }

  async getAllProfiles() {
    const cursor = await rdb.table('profiles').run(this.conn)
    const data = await cursor.toArray()
    await cursor.close()
    return data
  }

  async reconnect() {
    this.conn = await this.conn.reconnect({noreplyWait: false})
    conn.on('close', this.onClose.bind(this));
    conn.on('timeout', this.onTimeout.bind(this));
  }

  onError(err) {
    console.log('[DB]', 'Error', err)
    this.emit('error', err)
  }

  async onClose() {
    console.log('[DB]', 'Closed')
    this.emit('close')
    await this.reconnect()
  }

  async onTimeout() {
    console.log('[DB]', 'Connection Timeout')
    this.emit('timeout')
    await this.reconnect()
  }
}


routes/404.js
/*
 This file is part of Turquoise.
 Copyright (c) Snazzah 2016-2019
 Copyright (c) Yamboy1 (and contributors) 2019

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const router = require('express').Router();

router.get('/*', (req, res, next) => {
    res.displayError();
});

module.exports = router;


routes/connect.js
/*
 This file is part of Turquoise.
 Copyright (c) Snazzah 2016-2019
 Copyright (c) Yamboy1 (and contributors) 2019

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const express = require('express')
const btoa = require('btoa')
const sf = require('snekfetch')
const router = express.Router()
const config = require('../config.json')

router.get('/:id', async (req, res) => {
  if(!req.params.id) return res.displayError(400, 'Empty Profile ID')
  let profile = await req.db.getProfile(req.params.id)
  if(!profile) return res.displayError(400, 'Invalid Profile ID')
  if(!req.query.code){
    res.renderVue('connecting',
      { profile: profile.id, profileName: profile.name, fetching: true, match: profile.match, token: '', ticket: '' },
      { head: { meta: [ { script: '/js/connect.js' } ] } })
  } else {
    const auth = btoa(`${profile.client_id}:${profile.secret}`)
    let redirect = encodeURIComponent(`${config.ip}/connect/${profile.id}`)
    try {
      let tokenres = await sf.post(`https://discordapp.com/api/oauth2/token?grant_type=authorization_code&code=${req.query.code}&redirect_uri=${redirect}`)
        .set('Authorization', `Basic ${auth}`)
      console.log(tokenres.body)

      if(tokenres.body.scope !== 'identify guilds') return res.displayError(400, 'Invalid Scope')

      let userres = await sf.get(`https://discordapp.com/api/users/@me`)
        .set('Authorization', `Bearer ${tokenres.body.access_token}`)

      let ticket = req.tokengen.generate()
      await req.db.addTicket(ticket, profile.id, tokenres.body.access_token, userres.body)

      res.renderVue('connecting',
        { profile: profile.id, profileName: profile.name, fetching: false, match: profile.match, token: tokenres.body.access_token, ticket },
        { head: { meta: [ { script: '/js/connect.js' } ] } })
    } catch (e) {
      res.displayError(500, e.toString())
      console.error(e)
    }
  }
  //res.send({id:req.params.id, profile})
  //res.redirect(`https://discordapp.com/oauth2/authorize?client_id=${CLIENT_ID}&scope=identify&response_type=code&redirect_uri=${redirect}`);
})


router.get('/:id/redirect', async (req, res) => {
  if(!req.params.id) return res.displayError(400, 'Empty Profile ID')
  let profile = await req.db.getProfile(req.params.id)
  if(!profile) return res.displayError(400, 'Invalid Profile ID')
  let redirect = encodeURIComponent(`${config.ip}/connect/${profile.id}`)
  res.redirect(`https://discordapp.com/oauth2/authorize?client_id=${profile.client_id}&scope=identify%20guilds&response_type=code&redirect_uri=${redirect}`);
})

module.exports = router


routes/endpoints.js
/*
 This file is part of Turquoise.
 Copyright (c) Snazzah 2016-2019
 Copyright (c) Yamboy1 (and contributors) 2019

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const express = require('express')
const btoa = require('btoa')
const sf = require('snekfetch')
const router = express.Router()

router.get('/me', (req, res) => res.redirect('https://discordapp.com/api/users/@me'))
router.get('/guilds', (req, res) => res.redirect('https://discordapp.com/api/users/@me/guilds'))
router.get('/connections', (req, res) => res.redirect('https://discordapp.com/api/users/@me/connections'))

module.exports = router


public/trello/auth/script.js
/*
 This file is part of Turquoise.
 Copyright (c) Snazzah ???-2019
 Copyright (c) Yamboy1 (and contributors) 2019

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const qs = (function(a) {
  if (a == "") return {};
  var b = {};
  for (var i = 0; i < a.length; ++i) {
    var p=a[i].split('=', 2);
    if (p.length == 1)
      b[p[0]] = "";
    else
      b[p[0]] = decodeURIComponent(p[1].replace(/\+/g, " "));
  }
  return b;
})(window.location.search.substr(1).split('&'))
history.pushState(null, "", location.href.split("?")[0])

if(qs.t && qs.tk) {
  localStorage.setItem('token', qs.t)
  localStorage.setItem('ticket', qs.tk)
}

let token = qs.t || localStorage.token || '';
let ticket = qs.tk || localStorage.ticket || '';
if(!token || !ticket) window.location.href = `https://auth.trellobot.xyz/connect/trello#${window.location.href.replace(/((#|\?).+)/g, '')}`
if(!localStorage.token) localStorage.token = token
if(!localStorage.ticket) localStorage.ticket = ticket

window.discordRequest = async e => {
  try {
    let r = await fetch(`https://discordapp.com/api/users/${e}`, { headers: {
      'authorization': `Bearer ${token}`
    }})
    let json = await r.json()
    return json
  } catch (e) {
    localStorage.removeItem('token')
    localStorage.removeItem('ticket')
    window.location.href = `https://auth.trellobot.xyz/connect/trello#${window.location.href.replace(/((#|\?).+)/g, '')}`
  }
}

window.logout = () => {
  localStorage.removeItem('token')
  localStorage.removeItem('ticket')
  window.location.href = `https://auth.trellobot.xyz/connect/trello#${window.location.href.replace(/((#|\?).+)/g, '')}`
}

window.onload = () => {
  console.log('loaded')
  Trello.authorize({
    persist: false,
    scope: {
      read: true,
      write: true,
      account: true
    },
    expiration: 'never',
    success: async () => {
      console.log('yay we got things done!')
      console.log(Trello.token());
      try{
        let res = await Snekfetch.get(`https://auth.trellobot.xyz/trello/post/${Trello.token()}`).set('Authorization', localStorage.ticket)
        console.log(res)
        document.querySelector('h1').innerText = "Hell yeah! It's done!"
      } catch (e) {
        if(e.statusCode === 401) return window.logout();
        if (e.statusCode === 400) return document.querySelector('h1').innerText = `Error: ${e.body}!!`
        console.log({e})
        document.querySelector('h1').innerText = "Oh no! Check console!"
      }
    },
    error: console.log,
    name: 'Trello Discord Bot'
  })
}


routes/trello.js
/*
 This file is part of Turquoise.

 Copyright © Snazzah 2016 - 2019
 Copyright © Yamboy1 (and contributors) 2019
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

const express = require("express");
const router = express.Router();
const config = require("../config.json");

router.get("/auth/", async (req, res) => {
  res.type('.html').send(`
    <!doctype html>
    <html>
      <head>
        <meta charset="utf-8"> 
        <meta http-equiv="X-UA-Compatible" content="chrome=1">
    
        <title>Trello Bot Authorization</title>
        <link href="https://fonts.googleapis.com/css?family=Ubuntu" rel="stylesheet">
        <link href="/trello/auth/style.css" type="text/css" rel="stylesheet">
        <script src="https://cdn.jsdelivr.net/npm/snekfetch@4.0.4/browser.min.js"></script>
        <script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
        <script src="https://trello.com/1/client.js?key=${config.trelloKey}"></script>
        <script src="/trello/auth/script.js"></script>
      </head>
      <body>
        <div class="main">
          <h1 class="mainbox">Authorizing...</h1>
        </div>
      </body>
    </html>
`);
});

router.get('/auth/script.js', (req, res) => {
  res.type('.js');
  res.send(`
  const qs = (function(a) {
    if (a == "") return {};
    var b = {};
    for (var i = 0; i < a.length; ++i) {
      var p=a[i].split('=', 2);
      if (p.length == 1)
        b[p[0]] = "";
      else
        b[p[0]] = decodeURIComponent(p[1].replace(/\\+/g, " "));
    }
    return b;
  })(window.location.search.substr(1).split('&'))
  history.pushState(null, "", location.href.split("?")[0])

  if(qs.t && qs.tk) {
    localStorage.setItem('token', qs.t)
    localStorage.setItem('ticket', qs.tk)
  }

  let token = qs.t || localStorage.token || '';
  let ticket = qs.tk || localStorage.ticket || '';
  if(!token || !ticket) window.location.href = \`${config.ip}/connect/trello#\${window.location.href.replace(/((#|\\?).+)/g, '')}\`
  if(!localStorage.token) localStorage.token = token
  if(!localStorage.ticket) localStorage.ticket = ticket

  window.discordRequest = async e => {
    try {
      let r = await fetch(\`https://discordapp.com/api/users/\${e}\`, { headers: {
          'authorization': \`Bearer \${token}\`
        }})
      let json = await r.json()
      return json
    } catch (e) {
      localStorage.removeItem('token')
      localStorage.removeItem('ticket')
      window.location.href = \`${config.ip}/connect/trello#\${window.location.href.replace(/((#|\\?).+)/g, '')}\`
    }
  }

  window.logout = () => {
    localStorage.removeItem('token')
    localStorage.removeItem('ticket')
    window.location.href = \`${config.ip}/connect/trello#\${window.location.href.replace(/((#|\\?).+)/g, '')}\`
  }

  window.onload = () => {
    console.log('loaded')
    Trello.authorize({
      persist: false,
      scope: {
        read: true,
        write: true,
        account: true
      },
      expiration: 'never',
      success: async () => {
        console.log('yay we got things done!')
        console.log(Trello.token());
        try{
          let res = await Snekfetch.get(\`${config.ip}/trello/post/\${Trello.token()}\`).set('Authorization', localStorage.ticket)
          console.log(res)
          document.querySelector('h1').innerText = "Hell yeah! It's done!"
        } catch (e) {
          if(e.statusCode === 401) return window.logout();
          if (e.statusCode === 400) return document.querySelector('h1').innerText = \`Error: \${e.body}!!\`
          console.log({e})
          document.querySelector('h1').innerText = "Oh no! Check console!"
        }
      },
      error: console.log,
      name: 'Trello Discord Bot'
    })
  }
  `);
});

router.get('/auth/style.css', (req,res) => {
  res.type('.css');
  res.send(`
  html, body {
  margin: 0;
  padding: 0;
  height: 100%;
  width: 100%;
  background-color: #1587ea;
  overflow: hidden;
}

body {
  font: 14px/1.5 Ubuntu, Lato, "Helvetica Neue", Helvetica, Arial, sans-serif;
  font-weight: 300;
  color: white;
  -webkit-app-region: drag;
  -webkit-user-select: none;
  display: table;
  overflow-wrap: break-word;
  background-size: cover;
  background-position-x: center;
  background-position-x: center;
}

* {
  box-sizing: border-box;
}

.main {
  display: table-cell;
  vertical-align: middle;
  text-align: center;
}

.mainbox {
  vertical-align: middle;
  border-radius: 8px;
  background-color: rgba(0, 0, 0, .5);
  display: inline-block;
  width: 600px;
}
  `)
});

module.exports = router;
